File: Lab4/Queue.kt
import java.text.DecimalFormat

data class Queue(val maxSize: Int){
    val values = arrayOfNulls<Double>(maxSize)
    var front = 0
    var rear = -1
    var counter = 0
}

fun createQueue(size: Int): Queue?{
    if (size <= 0){
        println("Please use valid size which is bigger than 0")
        return null
    }
    return Queue(size)
}

fun isEmpty(queue: Queue): Boolean{
    return queue.counter == 0
}

fun isFull(queue: Queue): Boolean {
    return queue.counter >= queue.maxSize
}

fun enqueue(queue: Queue, x: Double): Boolean{
    if (isFull(queue)){
        println("Error: The queue is full, enqueue failed.")
        return false
    }
    queue.rear = (queue.rear + 1) % queue.maxSize
    queue.values[queue.rear] = x
    queue.counter++
    return true
}

fun dequeue(queue: Queue): Double?{
    if (isEmpty(queue)){
        println("Error: The queue is empty, dequeue failed.")
        return queue.values[queue.front]
    }
    queue.front = (queue.front + 1) % queue.maxSize
    queue.counter--
    when (queue.front == 0){
        true -> return queue.values[queue.maxSize - 1]
        false -> return queue.values[queue.front - 1]
    }
}

fun displayQueue(queue: Queue){
    val df: DecimalFormat = DecimalFormat("#.##")//Format Decimal numbers
    System.out.printf("front -->%8s\n", df.format(queue.values[queue.front]))
    if (queue.front > queue.rear){
        for (i in (queue.front + 1) until queue.maxSize){
            System.out.printf("%17s\n", df.format(queue.values[i]))
        }
        for (i in 0 until queue.rear){
            System.out.printf("%17s\n", df.format(queue.values[i]))
        }
    }
    else{
        for (i in (queue.front + 1) until queue.rear){
            System.out.printf("%17s\n", df.format(queue.values[i]))
        }
    }
    System.out.printf("%17s%17s\n", df.format(queue.values[queue.rear]), "<-- rear")
}

fun destroyQueue(queue: Queue){
    for (i in queue.values.indices){
        queue.values[i] = null
    }
}

File: Lab4/queue.cpp
#include <stdio.h>
#include <stdlib.h>
#include "queue.h"

bool CreateQueue(Queue *queue, int size) {//Create a new queue
	if (size <= 0)//Check if the size is valid
		return false;
	queue->values = (double *)malloc(sizeof(double) * size);//Alloc space for queue
	queue->front = 0;
	queue->rear = -1;
	queue->counter = 0;
	queue->maxSize = size;
	return true;
}

bool IsEmpty(Queue *queue) {//Check if the queue is empty
	return queue->counter == 0;
}

bool IsFull(Queue *queue) {//Check if the queue is full
	return queue->counter >= queue->maxSize;
}

bool Enqueue(Queue *queue, double x) {//Add element into the queue
	if (IsFull(queue))//If the queue is full, stop adding elements into it
		return false;
	queue->rear = (queue->rear + 1) % queue->maxSize;//Find the position of queue's tail
	queue->values[queue->rear] = x;
	queue->counter++;
	return true;
}

bool Dequeue(Queue *queue, double *x) {//Delete the first element in the queue
	if (IsEmpty(queue))//If the queue is empty, stop deleting elements in it
		return false;
	queue->front = (queue->front + 1) % queue->maxSize;//Find the position of queue's front
	if (queue->front == 0)
		*x = queue->values[queue->maxSize - 1];
	else
		*x = queue->values[queue->front - 1];
	queue->counter--;
	return true;
}

void DisplayQueue(Queue *queue) {//Display the queue
	if (IsEmpty(queue)){
		puts("The queue is empty.");
		return;
	}
	printf("front -->%8g\n", queue->values[queue->front]);
	if (queue->front > queue->rear) {
		for (int i = queue->front + 1; i < queue->maxSize; i++) {
			printf("%17g\n", queue->values[i]);
		}
		for (int i = 0; i < queue->rear; i++) {
			printf("%17g\n", queue->values[i]);
		}
	}
	else {
		for (int i = queue->front + 1; i < queue->rear; i++) {
			printf("%17g\n", queue->values[i]);
		}
	}
	printf("%17g%17s\n", queue->values[queue->rear], "<-- rear");
}

void DestroyQueue(Queue *queue) {
	free(queue->values);
}